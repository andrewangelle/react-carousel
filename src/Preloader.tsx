import React, { ReactElement } from 'react';

export interface PreloaderProps {
  images: string[];
}

export interface PreloaderState {
  loading: boolean;
  done: boolean;
}

export class Preloader extends React.Component<PreloaderProps, PreloaderState> {
  state: PreloaderState = {
    loading: true,
    done: false
  }

  componentDidMount() {
    this.preloadImages()
  }

  urls = () => {
    let urlResults: string[] = []

    React.Children.map(this.props.children,
      child => {
        const res = this.getImageUrls(child as ReactElement<any>)
        urlResults = [...urlResults, ...res]
      }
    )

    return urlResults
  }

  preloadImages() {
    const urls = [...this.props.images, ...this.urls()]
    let newImage
    let loadCount = 0

    const handleImageLoad = () => {
      loadCount = loadCount + 1

      if (loadCount === urls.length) {
        this.updateImageState();
      }
    }

    if (urls.length > 0) {
      urls.forEach(child => {
        console.log(child)
        newImage = new Image();
        newImage.onload = handleImageLoad;
        newImage.src = child;
      });
    }

    if (urls.length == 0) {
      this.updateImageState();
    }
  }

  getImageUrls(parent: ReactElement<any>) {
    const results: string[] = []
    const traverseTree = (object: any) => {
      if (object.type === 'img') {
        results.push(object.props.src)
      }
      if (object.props.children) {
        traverseTree(object.props.children)
      }
      return results
    }

    const res = traverseTree(parent)
    console.log({ res })
    return res
  }

  updateImageState() {
    this.setState({
      done: true,
      loading: false,
    });
  }
  render() {
    return (
      <>
        {this.state.loading && 'Loading ....'}
        {this.state.done && this.props.children}
      </>
    )
  }
}