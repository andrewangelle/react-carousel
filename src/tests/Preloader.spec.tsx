import React from 'react';
import { shallow } from './setupTests';
import { Preloader } from '../Preloader';

describe('Preloader', () => {
  afterEach(() => jest.clearAllMocks());
  it('shows loader', () => {
    const wrapper = shallow(
      <Preloader
        currentIndex={0}
        width={500}
        autoPlay={false}
        startTimer={jest.fn()}
      >
        <img src="https://picsum.photos/700" />
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Preloader>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('shows children', () => {
    const wrapper = shallow(
      <Preloader
        currentIndex={0}
        width={500}
        autoPlay={false}
        startTimer={jest.fn()}
      >
        <img src="https://picsum.photos/700" />
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Preloader>
    );

    wrapper.setState({ loading: false, done: true });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('fires timer cb after preloading is finished', () => {
    const props = {
      currentIndex: 0,
      width: 500,
      autoPlay: true,
      startTimer: jest.fn()
    };
    const wrapper = shallow(
      <Preloader {...props}>
        <img src="https://picsum.photos/700" />
        <h1>Slide Two</h1>
        <h1>Slide Three</h1>
      </Preloader>
    );
    expect(props.startTimer).not.toHaveBeenCalled();

    wrapper.setState({ loading: false, done: true });
    wrapper.update();
    expect(props.startTimer).toHaveBeenCalled();
  });
});
