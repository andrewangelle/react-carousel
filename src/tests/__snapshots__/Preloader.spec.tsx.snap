// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preloader shows children 1`] = `
<Fragment>
  <img
    className="glide--item current"
    key=".0"
    src="https://picsum.photos/700"
  />
  <h1
    className="glide--item "
    key=".1"
  >
    Slide Two
  </h1>
  <h1
    className="glide--item "
    key=".2"
  >
    Slide Three
  </h1>
</Fragment>
`;

exports[`Preloader shows loader 1`] = `
<Fragment>
  <LoadingSpinner
    width={500}
  />
</Fragment>
`;
